llvm/lib/Target/DirectX/DXIL.td

```diff
@@ -7,138 +7,139 @@
-  list<dxil_param> ops = []; // the operands that this instruction takes
-  string oload_types = "";   // overload types if applicable
-  string fn_attr = "";       // attribute shorthands: rn=does not access
-                             // memory,ro=only reads from memory,
-  bit is_deriv = 0;          // whether this is some kind of derivative
-  bit is_gradient = 0;       // whether this requires a gradient calculation
-  bit is_feedback = 0;       // whether this is a sampler feedback op
-  bit is_wave = 0; // whether this requires in-wave, cross-lane functionality
-  bit requires_uniform_inputs = 0; // whether this operation requires that all
-                                   // of its inputs are uniform across the wave
-  // Group dxil operation for stats.
-  // Like how many atomic/float/uint/int/... instructions used in the program.
-  list<string> stats_group = [];
+// A representation for a DXIL operation
+class DXILOperationDesc<string name> {
+  // TODO : Appears redundant. OpName should serve the same purpose
+  string Name = name; // short, unique name
+
+  string OpName = "";         // Name of DXIL operation
+  int OpCode = 0;             // Unique non-negative integer associated with the operation
+  DXILOpClass  OpClass;       // Class of the operation
+  DXILOpCategory OpCategory;  // Category of the operation
+  string Doc = "";            // Description of the operation
+  list<DXILOpParameter> Params = []; // Parameter list of the operation
+  string OverloadTypes = "";  // Overload types, if applicable
+  string Attributes = "";     // Attribute shorthands: rn=does not access
+                              // memory,ro=only reads from memory,
+  bit IsDerivative = 0;       // Whether this is some kind of derivative
+  bit IsGradient = 0;         // Whether this requires a gradient calculation
+  bit IsFeedback = 0;         // Whether this is a sampler feedback operation
+  bit IsWave = 0;             // Whether this requires in-wave, cross-lane functionality
+  bit NeedsUniformInputs = 0; // Whether this operation requires that all
+                              // of its inputs are uniform across the wave
+  // Group DXIL operation for stats - e.g., to accumulate the number of atomic/float/uint/int/...
+  // operations used in the program.
+  list<string> StatsGroup = [];
 }
 
-class dxil_op<string name, int code_id, dxil_class code_class, dxil_category op_category, string _doc,
-              string _oload_types, string _fn_attr, list<dxil_param> op_params,
-              list<string> _stats_group = []> : dxil_inst<name> {
-  let dxil_op = name;
-  let dxil_opid = code_id;
-  let doc = _doc;
-  let ops = op_params;
-  let op_class = code_class;
-  let category = op_category;
-  let oload_types = _oload_types;
-  let fn_attr = _fn_attr;
-  let stats_group = _stats_group;
+class DXILOperation<string name, int opCode, DXILOpClass opClass, DXILOpCategory opCategory, string doc,
```

Comment:
It might be too late given uses in other places, but I think using CamlelCase for `Dxil` instead of all caps `DXIL` (e.g. `DxilOperation<...>`) would be nicer to read.

Reply:
> It might be too late given uses in other places, but I think using CamlelCase for `Dxil` instead of all caps `DXIL` (e.g. `DxilOperation<...>`) would be nicer to read.

I certainly considered that - but argued against myself to stick with what did appear to be the usage in other places, as you point out. Additionally, I convinced myself that DXIL is (almost) an acronym for DirectX Intermediate Language (correct?) and hence, CamelCase may not be apt. That said, I remain open to other perspectives on this.

Reply:
Elsewhere (in DXC), we do have an established pattern of using `Dxil` when in CamelCase with other words, and `DXIL` when used by itself.  Perhaps we can establish the preferred pattern here before we have too much code depending on one pattern or the other.

Reply:
> Elsewhere (in DXC), we do have an established pattern of using `Dxil` when in CamelCase with other words, and `DXIL` when used by itself. Perhaps we can establish the preferred pattern here before we have too much code depending on one pattern or the other.

It is good to know of the prevailing established pattern. Thanks!

I'll update the PR to follow the same convention in `DXIL.td`.